### PR TITLE
nixos/modemmanager: support additional FCC unlock scripts

### PR DESCRIPTION
**Pull request by flokli**
https://github.com/NixOS/nixpkgs/pull/155414 introduced an option to support enabling the FCC unlock scripts that ModemManager provides, but doesn't execute (since 1.18.4). However, it wasn't possible to add additional, provided-by-the-vendor unlock scripts.

This commit fixes this up, by renaming the option from `networking.networkmanager.enableFccUnlock` to
`networking.networkmanager.enableBundledFccUnlockScripts`, and adding an additional option `networking.networkmanager.fccUnlockScripts`, which allows configuring (other) unlock scripts.

The logic is used in case
`networking.networkmanager.enableBundledFccUnlockScripts` is set to true - it'll then pick up the list of exposed
modemmanager.passthru.fccUnlockScripts and set/extend `networking.networkmanager.fccUnlockScripts` with them.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @jwygoda

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2205.section.md [hunk 1]
```diff
@@ -936,9 +936,10 @@ In addition to numerous new and upgraded packages, this release has the followin
   package.
 
 - The option
-  [services.networking.networkmanager.enableFccUnlock](#opt-networking.networkmanager.enableFccUnlock)
-  was added to support FCC unlock procedures. Since release 1.18.4, the ModemManager
-  daemon no longer automatically performs the FCC unlock procedure by default. See
+  [services.networking.networkmanager.enableBundledFccUnlockScripts](#opt-networking.networkmanager.enableBundledFccUnlockScripts)
+  was added to support FCC unlock procedures.
+  Since release 1.18.4, the ModemManager daemon no longer automatically performs
+  the FCC unlock procedure by default. See
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2205.section.md:942]: We shouldn't modify old release notes.

File: nixos/modules/services/networking/networkmanager.nix [hunk 2]
```diff
@@ -510,6 +537,16 @@ in {
         ];
       }
 
+      # if cfg.enableBundledFccUnlockScripts is set, populate
+      # networking.networkmanager.fccUnlockScripts with the values from
+      # pkgs.modemmanager.passthru.fccUnlockScripts.
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/networkmanager.nix:542]: We don't have networking.networkmanager.package?

File: nixos/modules/services/networking/networkmanager.nix [hunk 3]
```diff
@@ -369,16 +369,38 @@ in {
...
-      enableFccUnlock = mkOption {
+      enableBundledFccUnlockScripts = mkOption {
         type = types.bool;
         default = false;
         description = lib.mdDoc ''
-          Enable FCC unlock procedures. Since release 1.18.4, the ModemManager daemon no longer
+          Enable FCC unlock procedures shipped with ModemManager.
+          Since release 1.18.4, the ModemManager daemon no longer
           automatically performs the FCC unlock procedure by default. See
           [the docs](https://modemmanager.org/docs/modemmanager/fcc-unlock/)
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/networkmanager.nix:379]:
```suggestion
          [the modemmanager docs](https://modemmanager.org/docs/modemmanager/fcc-unlock/)
```

File: nixos/doc/manual/release-notes/rl-2205.section.md (hunk 1 above)
Comment (jtojnar) [reply nixos/doc/manual/release-notes/rl-2205.section.md:942]: This is required in order not to break link checker.

File: pkgs/tools/networking/modemmanager/default.nix [hunk 4]
```diff
@@ -93,6 +94,21 @@ stdenv.mkDerivation rec {
...
+    # Most of them actually symlink to a "common" unlock script
+    fccUnlockScripts = {
+      "03f0:4e1d" = "${modemmanager}/share/ModemManager/fcc-unlock.available.d/1199";
+      "105b:e0ab" = "${modemmanager}/share/ModemManager/fcc-unlock.available.d/105b";
+      "1199:9079" = "${modemmanager}/share/ModemManager/fcc-unlock.available.d/1199";
+      "1eac:1001" = "${modemmanager}/share/ModemManager/fcc-unlock.available.d/1eac";
+      "2c7c:030a" = "${modemmanager}/share/ModemManager/fcc-unlock.available.d/2c7c";
+      "413c:81a3" = "${modemmanager}/share/ModemManager/fcc-unlock.available.d/1199";
+      "413c:81a8" = "${modemmanager}/share/ModemManager/fcc-unlock.available.d/1199";
+    };
```
Comment (jtojnar) [inline pkgs/tools/networking/modemmanager/default.nix:109]: Please add a check to `passthru.tests` ensuring the code is kept in sync (see e. g. NetworkManager or fwupd packages).

File: nixos/modules/services/networking/networkmanager.nix [hunk 5]
```diff
@@ -387,7 +409,8 @@ in {
       [ "networking" "networkmanager" "packages" ]
       [ "networking" "networkmanager" "plugins" ])
     (mkRenamedOptionModule [ "networking" "networkmanager" "useDnsmasq" ] [ "networking" "networkmanager" "dns" ])
-    (mkRemovedOptionModule ["networking" "networkmanager" "dynamicHosts"] ''
+    (mkRenamedOptionModule [ "networking" "networkmanager" "enableFccUnlock" ] [ "networking" "networkmanager" "enableBundledFccUnlockScripts" ])
+    (mkRemovedOptionModule [ "networking" "networkmanager" "dynamicHosts" ] ''
```
Comment (jtojnar) [inline nixos/modules/services/networking/networkmanager.nix:413]: Could you please extract the cleanups to a separate commit?

File: nixos/modules/services/networking/networkmanager.nix (hunk 2 above)
Comment (flokli) [reply nixos/modules/services/networking/networkmanager.nix:542]: The unlock scripts are provided by modemmanager, not NetworkManager, even though it's in the "networkmanager" config namespace.

Moving this into a separate namespace is out of scope for this PR.

File: nixos/modules/services/networking/networkmanager.nix (hunk 5 above)
Comment (flokli) [reply nixos/modules/services/networking/networkmanager.nix:413]: I can split the rename into a commit before the actual change, sure.

File: pkgs/tools/networking/modemmanager/default.nix (hunk 4 above)
Comment (flokli) [reply pkgs/tools/networking/modemmanager/default.nix:109]: I don't think modemmanager will include more unlock scripts ever. It looks like that vendors will provide their own custom unlock scripts, rather than playing this whack-a-mole game, as least when reading https://modemmanager.org/docs/modemmanager/fcc-unlock/, which also lists these as an (exhaustive) list.

I'd personally prefer dropping all of `enableBundledFccUnlockScripts`, and leave it up to the user to explicitly set their unlock script (in case more SAR-aware scripts pop up somewhere) over trying to make this update-proof. I only kept this for backwards-compat with the previous NixOS option.

File: nixos/modules/services/networking/networkmanager.nix (hunk 3 above)
Comment (flokli) [reply nixos/modules/services/networking/networkmanager.nix:379]: This is unrelated to this PR and was already present before.

File: nixos/doc/manual/release-notes/rl-2205.section.md (hunk 1 above)
Comment (flokli) [reply nixos/doc/manual/release-notes/rl-2205.section.md]: I removed the hyperlink, put it in backticks, and squashed the two lines together.

File: nixos/modules/services/networking/networkmanager.nix (hunk 5 above)
Comment (flokli) [reply nixos/modules/services/networking/networkmanager.nix:413]: pushed.

**Comment by flokli** [comment]
@jtojnar poke ;-)

As written previously, I think having the user explicitly choose which unlock script to use (so the `networking.networkmanager.fccUnlockScripts = [{id = "vid:pid"; path = "…"}]` path) over the "enable all unlock scripts known route.

We could also drop the `enableBundledFccUnlockScripts` option alltogether, and explain how to add modemmanager-provided unlock scripts to the release notes. 

This is more in line with the upstream documentation, where users expliticly do symlinks for a pid:vid of their choosing, if they decide so.

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 252283` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>

File: pkgs/tools/networking/modemmanager/default.nix (hunk 4 above)
Comment (RaitoBezarius) [reply pkgs/tools/networking/modemmanager/default.nix:109]: I agree with @flokli on this one, @jtojnar do you think this is a blocking matter?

Comment (flokli) [reply pkgs/tools/networking/modemmanager/default.nix:109]: Just to make sure, my /preference/ would be to entirely drop the auto enabling of the bundled FCC unlock scripts, and document how to enable the bundled ones in the release notes. I'm happy to update the PR to that, but we need to decide which direction to go.

Comment (jtojnar) [reply pkgs/tools/networking/modemmanager/default.nix:109]: Sorry, I thought I have already replied that I am fine with removing the `enableBundledFccUnlockScripts` but apparently not.

Comment (flokli) [reply pkgs/tools/networking/modemmanager/default.nix:109]: No problem, thanks for confirming. I'll update the PR.

**Comment by flokli** [comment]
As discussed in https://github.com/NixOS/nixpkgs/pull/252283#discussion_r1323418885, I updated the PR to now remove the networking.networkmanager.enableFccUnlock script entirely. PTAL.

**Review by RaitoBezarius — APPROVED** [review]
